### ISSUE TITLE
RL - DQN & FlappyBird Implementation

### ISSUE THREAD

**Issue by Seryol** [Seryol/Colythme#12]
[A]: # 'https://s1.ax1x.com/2022/06/22/jptiv9.jpg'

>Everywhere&#x0020;in&#x0020;our&#x0020;life,&#x0020;there&#x0020;exists&#x0020;philosophy.<br><br>在训练这只笨鸟的时候，我看透了人生

First of all, let me give a basic intuition of `$DQN$` (Deep Q-Network). It is composed of `$D$` (deep learning) and `$QN$` (Q-Network), among which the Q-Learning algorithm is the core and the neural network is just a tool to maintain the matrix in Q algorithm because of the memory limits (a `$state \times actioin$` size of matrix is too big for most of the times).

### Q-Learning

It is explained in detail in this blog: [[A Painless Q-Learning Tutorial](http://mnemstudio.org/path-finding-q-learning-tutorial.htm)](http://mnemstudio.org/path-finding-q-learning-tutorial.htm)

### DQN

Similar to dynamic programming, the state transition equation of Q array presented in Q-Learning:
`$$
Q (s, a) = R(s, a) + \gamma \cdot \max_{\tilde{a}}\{Q (\tilde{s}, \tilde{a})\}
$$`
Here `$s, a$` indicates current state and action, `$R$` indicates the reward matrix, `$\gamma$` indicates learning rate (a coefficient), `$\tilde{s}, \tilde{a}$` indicates state and action in the next time step.

Then comes the problem: how do we measure loss function? Easy. Remind that we are now using neural network to maintain the Q-Learning sheet, which means we need to have our network's outputs approximate to Q values. Therefore, it can now be regarded as supervised learning with labels as Q values calculated by previous networks.
`$$
L_i(\theta_i) = (r + \gamma \cdot (\max_{\tilde{a}}\{Q(\tilde{s}, \tilde{a} | \theta_{i - 1})\}) - Q(s, a | \theta_i))^2
$$`
which is an MSE loss function, and `$\theta$` means we need to use parametric approximations.

The training process is divided into three parts: observation period, exploration period, and training period

In practice, we define a replay memory to save those existed states and randomly pick up a batch of them as what we are to train in the exploration and training period, which can make the training more efficient. Meanwhile, we need to build two same networks, Q-Network and target Q-Network. Q-Network is what we are training, but target Q-Network is used to predict 'label Q values'. And every certain episodes, we copy the parameters in Q-Network to target Q-Network.

## Code & Annotations

[Code reference](https://github.com/SukerZ/Playing-Flappy-Bird-by-DQN-on-PyTorch)

```python
import os
import sys
import cv2
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import random
from collections import deque

# import game (pygame)
sys.path.append ('game/')
import wrapped_flappy_bird as GameMain

# collect data only in OBSERVE_PERIOD
# train the network and process annealing to epsilon in EXPLORE_PERIOD
# then comes the final period, epsilon reaches its ultimate value, it'll not be changed any more
OBSERVE_PERIOD = 1000
EXPLORE_PERIOD = 2000000
UPDATE_TIME = 100 # update Q_NetT with Q_Net parameters every UPDATE_TIME time steps

ACTIONS = 2 # number of actions
ACTION_PER_STEP = 1 # execute an action every # time steps
GAMMA = 0.99
EPSILON_INIT = 0.1 # initial value of epsilon
EPSILON_ULTIMATE = 0.0001 # the ultimate value of epsilon after declining

MAX_MEMORY = 50000 # maximum elements saved in the replay memory

BATCH_SIZE = 32
LEARNING_RATE = 1e-06

def image_handler (img):
    img = cv2.cvtColor (cv2.resize (img, (80, 80)), cv2.COLOR_BGR2GRAY)
    # binarization: threshold is 1 (note that the original background is totally black instead of scene)
    _, img = cv2.threshold (img, 1, 255, cv2.THRESH_BINARY)
    return img

# output_size = (input_size - kernel_size + 2 * padding) / stride + 1
class DQN_nn (nn.Module):
    def __init__ (self):
        super (DQN_nn, self).__init__ ()
        self.conv1 = nn.Sequential (
            nn.Conv2d (in_channels = 4, out_channels = 32, kernel_size = 8, stride = 4, padding = 2),
            nn.ReLU (),
            nn.MaxPool2d (kernel_size = 2)
        )
        self.conv2 = nn.Sequential (
            nn.Conv2d (in_channels = 32, out_channels = 64, kernel_size = 4, stride = 2, padding = 1),
            nn.ReLU ()
        )
        self.conv3 = nn.Sequential (
            nn.Conv2d (in_channels = 64, out_channels = 64, kernel_size = 3, stride = 1, padding = 1),
            nn.ReLU ()
        )
        self.fc1 = nn.Sequential (
            nn.Linear (1600, 512),
            nn.ReLU ()
        )
        self.fc2 = nn.Linear (512, ACTIONS)
    
    def forward (self, x):
        x = self.conv1 (x)
        x = self.conv2 (x)
        x = self.conv3 (x)
        x = x.view (x.size (0), - 1)
        x = self.fc1 (x)
        x = self.fc2 (x)
        return x

# inherit 'object' class is for compatibility
class _DQN_ (object):
    def __init__ (self, actions):
        self.actions = actions
        self.QNet = DQN_nn ()
        self.QNet_T = DQN_nn ()
        if os.path.exists ('model/fb_qnet.pt'):
            premodel = torch.load ('model/fb_qnet.pt')
            self.QNet.load_state_dict (premodel)
            self.QNet_T.load_state_dict (premodel)
        self.timesteps = 0
        self.epsilon = EPSILON_INIT
        self.replayMemory = deque ()

    def state_init (self, observation):
        # np.stack with axis = 0 can 'pile up' the arrays
        # now self.state.shape should be 80 * 80 * 4
        self.state = np.stack ((observation, observation, observation, observation), axis = 0)

    # return next action
    def step_action (self):
        QValue = self.QNet (torch.tensor ([self.state]).to (torch.float32))[0]
        action = np.zeros (self.actions)
        if self.timesteps % ACTION_PER_STEP == 0:
            if random.random () <= self.epsilon: # picks a random action
                exe_action = random.randrange (self.actions)
                action[exe_action] = 1
            else: # picks an action according to QValue
                # use detach() before swithing from torch tensor to numpy array
                # so that there'll not be gradients
                exe_action = np.argmax (QValue.detach ().numpy ())
                action[exe_action] = 1
        else:
            action[0] = 1
        
        # annealing
        if self.timesteps > OBSERVE_PERIOD and self.epsilon > EPSILON_ULTIMATE:
            self.epsilon -= (EPSILON_INIT - EPSILON_ULTIMATE) / EXPLORE_PERIOD
        
        return action
    
    def train (self):
        optimizer = torch.optim.Adam (self.QNet.parameters (), lr = LEARNING_RATE)

        # random.sample(array, n) can randomly pick n elements from array and returns an array
        minibatch = random.sample (self.replayMemory, BATCH_SIZE)
        state_batch = [items[0] for items in minibatch]
        action_batch = [items[1] for items in minibatch]
        new_state_batch = [items[2] for items in minibatch]
        reward_batch = [items[3] for items in minibatch]
        done_batch = [items[4] for items in minibatch]

        # target_Q is used for saving Q values calculated in target netwrok
        target_Q = np.zeros ([BATCH_SIZE, 1])
        new_state_batch = torch.tensor (np.array (new_state_batch)).to (torch.float32)
        tar_QVal = self.QNet_T (new_state_batch)
        tar_QVal = tar_QVal.detach ().numpy ()
        for i in range (BATCH_SIZE):
            cdone = minibatch[i][4]
            if cdone == True:
                target_Q[i][0] = reward_batch[i]
            else:
                # note that Q_NetT use max function but Q_Net's result only depends on actions
                # for target_Q is deducted with Q-Learning process
                target_Q[i][0] = reward_batch[i] + GAMMA * np.max (tar_QVal[i])
        target_Q = torch.tensor (np.reshape (np.array (target_Q), [BATCH_SIZE, 1]))
        # eval_Q is used for saving Q values calculated in Q_Net
        eval_Q = self.QNet (torch.tensor (state_batch).to (torch.float32))
        # before reshape the shape is (BATCH_SIZE, ), therefore, we need to reshape
        index = torch.tensor (np.reshape (np.argmax (np.array (action_batch), axis = 1), [BATCH_SIZE, 1]))
        eval_Q = eval_Q.gather (1, index)
        # gather function in torch:
        # ------------------------------
        # e.g. a = [[1, 2, 3], [4, 5, 6]]
        # b = [[1, 2], [0, 1]]
        # a.gather (axis = 1, b) = [[2, 3], [4, 5]]
        # which means we pick elements in a with corresponding indices given in b
        # ------------------------------

        loss = F.mse_loss (eval_Q.to (torch.float32), target_Q.to (torch.float32))
        optimizer.zero_grad ()
        loss.backward ()
        optimizer.step ()

        if self.timesteps % UPDATE_TIME == 0:
            self.QNet_T.load_state_dict (self.QNet.state_dict ())
            torch.save (self.QNet.state_dict (), 'model/fb_qnet.pt')
            print ("save model once")
    
    def update_memory_replay (self, action, observation, reward, done):
        # now the new state should be
        # [last_observation, last_observation, last_observation, new_observatioin]
        new_state = np.stack ((self.state[1, :, :], self.state[2, :, :], self.state[3, :, :], observation), axis = 0)
        self.replayMemory.append ((self.state, action, new_state, reward, done))
        if len (self.replayMemory) > MAX_MEMORY:
            self.replayMemory.popleft ()
        if self.timesteps > OBSERVE_PERIOD:
            self.train ()
        
        state = ""
        if self.timesteps <= OBSERVE_PERIOD:
            state = "Observation"
        elif self.timesteps <= OBSERVE_PERIOD + EXPLORE_PERIOD:
            state = "Explore"
        else:
            state = "Train"
        print ("TIMESTEP: ", self.timesteps, "/ STATE: ", state, "/ EPSILON:", self.epsilon)
        self.state = new_state
        self.timesteps += 1

if __name__ == "__main__":
    # GameState is the main class in FlappyBird pygame, which saves current state of the agent
    game = GameMain.GameState ()
    DQN = _DQN_ (ACTIONS)
    # action = [1, 0] means drop, [0, 1] means fly
    action = np.array ([1, 0])
    # frame_step function has the same effect as the step function in gym
    # observation is a frame of the game (image)
    observation, reward, done = game.frame_step (action)
    observation = image_handler (observation) # reshape & binarization
    DQN.state_init (observation)
    
    # Q-Learning:
    # acquire next action -> calculate Q -> send back to memory replay
    while True:
        action = DQN.step_action ()
        observation, reward, done = game.frame_step (action)
        observation = image_handler (observation)
        DQN.update_memory_replay (action, observation, reward, done)
```

## 一些qiqiguaiguai想法

看着这只笨鸟训练，我感慨万千

若没有初始导入预训练的模型，大部分时候它都飞不过第一根柱子

为什么呢

模型参数相当于知识，而这鸟就像是某个人的替身

他一开始没有足够的知识让他穿过第一根柱子（success），那么他就永远不会得到第一笔大的回报（reward）

他没成功过，那么他永远也不知道怎么去成功

当然了，不排除一些天选之子，上天给予了他机会让他穿过第一根柱子（`$1 - \gamma$`），或许接下来的柱子对他也不是难题

而那些没有被选中的，也只会，只能注重手头上的小利（survive: + 1 point）从而不断挣扎着往顶上飞，然后过了一会儿便撞上了柱子

嘛，不过老话说得好，量变到质变

我因为电脑没训练一会儿就太烫了所以只看了几千个time step，观测到了如此结果

但我设置的total time steps实际上得有几十万，或许积累到了那个量，形成了质变，这个笨鸟也懂得穿过第一根柱子了呢？
